eclipse-theia/theia
# Add URI.joinPath() API

#### What it does

This is another go at "vscode.Uri.joinPath is not a function" #8752.

The approach (like in the original PR from @svor) is to derive our URI implementation from the vscode-uri implementation we use elsewhere in Theia. In order for this to work, we do the following:

1. Create only instances of our own class inside the plugin host process
2. Whenever we transfer a vscode-uri into the plugin host process, we revive it as an instance of our own class
3. Whenever we transfer and instance of our own class outside of the plugin host process, we marshal it as a vscode-uri.

In order to achieve this, I've made the replacer/reviver in rpc-protocol.ts pluggable.

#### How to test

Same as https://github.com/eclipse-theia/theia/pull/9199, but make sure that the errors found post-merge are not occurring (in particular with the Timeline view.


#### Review checklist


#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
